network/transport/quic.go

```diff
@@ -0,0 +1,196 @@
+ *    limitations under the License.
+ */
+
+package transport
+
+import (
+	"context"
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/pem"
+	"math/big"
+	"net"
+
+	"github.com/insolar/insolar/log"
+	"github.com/insolar/insolar/network/transport/relay"
+	"github.com/lucas-clemente/quic-go"
+	"github.com/pkg/errors"
+)
+
+type quicConnection struct {
+	session quic.Session
+	stream  quic.Stream
+}
+
+type quicTransport struct {
+	baseTransport
+	l           quic.Listener
+	conn        net.PacketConn
+	connections map[string]quicConnection
+}
+
+func newQuicTransport(conn net.PacketConn, proxy relay.Proxy, publicAddress string) (*quicTransport, error) {
+	listener, err := quic.Listen(conn, generateTLSConfig(), nil)
+	if err != nil {
+		return nil, err
+	}
+
+	transport := &quicTransport{
+		baseTransport: newBaseTransport(proxy, publicAddress),
+		l:             listener,
+		conn:          conn,
+		connections:   make(map[string]quicConnection),
+	}
+
+	transport.sendFunc = transport.send
+	return transport, nil
+}
+
+func (q *quicTransport) send(recvAddress string, data []byte) error {
+	conn, ok := q.connections[recvAddress]
+	var stream quic.Stream
+	var err error
+	if !ok {
+		var session quic.Session
+		session, stream, err = createConnection(recvAddress)
+		if err != nil {
+			return errors.Wrap(err, "[ send ] failed to create a connection")
+		}
+		q.connections[recvAddress] = quicConnection{session, stream}
+	} else {
+		stream = conn.stream
+	}
+
+	n, err := stream.Write(data)
+	if err != nil {
+		return errors.Wrap(err, "[ send ] failed to write to a stream")
+	}
+
+	if n != len(data) {
+		return errors.New("[ send ] sent a part of data")
+	}
+
+	return nil
+}
+
+// Start starts networking.
+func (q *quicTransport) Start(ctx context.Context) error {
+	log.Debug("Start QUIC transport")
+	for {
+		session, err := q.l.Accept()
+		if err != nil {
+			<-q.disconnectFinished
+			return err
+		}
+
+		log.Debugf("accept from: %s", session.RemoteAddr().String())
+		go q.handleAcceptedConnection(session)
+	}
+}
+
+// Stop stops networking.
+func (q *quicTransport) Stop() {
+	q.mutex.Lock()
+	defer q.mutex.Unlock()
+
+	log.Debug("[ Stop ] Stop QUIC transport")
+	q.prepareDisconnect()
+
+	err := q.l.Close()
+	if err != nil {
+		log.Errorln("[ Stop ] Failed to close socket:", err.Error())
+	}
+
+	err = q.closeConnections()
+	if err != nil {
+		log.Error(err, "[ Stop ] failed to close sessions")
+	}
+	err = q.conn.Close()
+	if err != nil {
+		log.Error(err, "[ Stop ] failed to close a connection")
+	}
+}
+
+func (q *quicTransport) handleAcceptedConnection(session quic.Session) {
+	stream, err := session.AcceptStream()
+	if err != nil {
+		log.Error(err, "[ handleAcceptedConnection ] failed to get a stream")
+	}
+
+	msg, err := q.serializer.DeserializePacket(stream)
+	if err != nil {
+		log.Error(err, "[ handleAcceptedConnection ] failed to deserialize a packet")
+	}
+
+	q.handlePacket(msg)
```

Comment:
i think we should start this in goroutine

Reply:
done